### ISSUE TITLE
Is there a way to bulk-publish all existing published pages into the default locale?

### ISSUE THREAD

**Issue by muppsy007** [tractorcow-farm/silverstripe-fluent#572]
I've done a search through the issues and had a look at dev/tasks, but I don't see anything like this. I just wanted to ask before creating a new task myself.

Existing site, about 200 pages. We have added Fluent which is fine. But by default none of the existing pages are actually published into the default locale (because there is no default locale until you make one, I guess). I was wondering if there was an existing method of taking all currently published pages in the sitetree and adding them to the default locale? (Or a directed locale)

If not, I will likely make a new task. Is it something you would want a PR for?

**Comment by robbieaverill** [comment]
I definitely think there's some space to improve this process. A build task would be the easiest entry point, but perhaps even Fluent could have a service class to do this set up if you check a box when creating your first locale like "Migrate existing content to this locale" or something?

What do you think @tractorcow?

**Comment by tractorcow** [comment]
You can just turn off `frontend_publish_required` and the pages will appear in all locales.

**Comment by tractorcow** [comment]
See https://github.com/tractorcow-farm/silverstripe-fluent/blob/4.4/docs/en/scenarios.md#content-inheritance-for-the-frontend

**Comment by muppsy007** [comment]
OK. Not really a solution to this problem, as it just replaces manually publishing a sitetree in one locale with manually unpublishing a sitetree in one or more different locales. Existing sites *always* have a current locale and are adding the module to begin adding translations. Which will often never be all pages.

I'll do a Task for myself. Cheers.

**Comment by tractorcow** [comment]
Hi @muppsy007 happy to re-open this issue if you want to continue to explore the problem. It's possible I don't fully understand your specific use case.

Typically removing the publish requirement causes pages to always be published in all locales, which is the general solution for avoiding that requirement.

However, as I understand it, you want pages to be published only in the default locale, am I correct?

**Comment by muppsy007** [comment]
Hi @tractorcow. If this is a feature that isn't appropriate to your plans feel free to close it. Perhaps I'm misunderstanding what you mean too. Here is my scenario:

1. Existing site of 200 pages in the CMS. All in English.
2. Owner would like to offer multilingual pages so install Fluent and set up two Locales. en_NZ (default) and es_MX.
3. Since the entire site was originally english, owner would like that to just work as normal. All current published pages are automatically published on en_NZ locale. I believe this is where `frontend_publish_required` will do the trick, but not for the next step as I understand it:
4. Since es_MX is a new Locale an the owner needs to gradually translate pages, and some won't actually be used, we want all pages to NOT be published in this locale.

So unless I am mistaken, we currently have an all or nothing approach to this. Either no pages are published on any locale and we need to manually do that for en_NZ in CMS. Or all pages are published on all locales and we need to manually unpublish pages on es_MX.

**Comment by tractorcow** [comment]
Feels like a good case for a build task we could throw in? Probably with 200 items I would not put that into the CMS (It would just hang for ages).

Something like `dev/tasks/FluentPublishPages?locale=es_MX`. Does this sound about right?

**Comment by muppsy007** [comment]
Yeah that was the original idea. I was talking to @robbieaverill about some specifics over on https://github.com/tractorcow-farm/silverstripe-fluent/issues/575#issuecomment-552695453. I was hoping to get something done tomorrow which I could share if it helps. As it stands I think I understand how this works.

1. Get all published pages
2. FluentState::setLocale($paramLocale)
3. Foreach published page as $pageObj
4. $pageObj->publishRecursive();

Bit more to it no doubt, but that's where I was heading if no other process existed.

**Comment by muppsy007** [comment]
@tractorcow @robbieaverill . Not sure if it's of any use by I made my task and put it in a Gist. Pretty basic, could probably do with some batch management as 200 pages definitely starts to hit walls with execution timeouts. But I run CLI anyway. Seems to do the trick.

https://gist.github.com/muppsy007/9f41eaebe12f71ebb41fc071a814279e

**Comment by tractorcow** [comment]
Thanks @muppsy007 :D

**Comment by jakxnz** [comment]
@muppsy007 thanks for sharing your gist! I was relieved to find someone had shared their effort.

I started with yours but found I had a few differences in what I needed:
- I didn't want to recursively publish records in the tree of recursive ownership, in case some content was still in DRAFT
- I didn't want to add a version record as part of the migration
- I wanted to run it as a job (symbiote/silverstripe-queuedjobs)
- I wanted to be able to re-run the job for discrete fields/columns without impacting previously "migrated" data (for cases where an existing field was changed to become translatable)

So I ended up making:
https://gist.github.com/jakxnz/b0c5402585560c04f96ed3d39399e46e

Note: It migrates a test batch of 300 records in less than a few seconds.
